tests/test_purchase_free_of_charge.py

```diff
@@ -0,0 +1,143 @@
+import unittest
+from decimal import Decimal
+
+from proteus import Model
+from trytond.modules.account.tests.tools import (create_chart,
+                                                 create_fiscalyear,
+                                                 get_accounts)
+from trytond.modules.account_invoice.tests.tools import (
+    create_payment_term, set_fiscalyear_invoice_sequences)
+from trytond.modules.company.tests.tools import create_company, get_company
+from trytond.tests.test_tryton import drop_db
+from trytond.tests.tools import activate_modules, set_user
+
+
+class Test(unittest.TestCase):
+
+    def setUp(self):
+        drop_db()
+        super().setUp()
+
+    def tearDown(self):
+        drop_db()
+        super().tearDown()
+
+    def test(self):
+
+
+        # Activate modules
+        activate_modules('purchase_line_free_of_charge')
+
+        # Create company
+        _ = create_company()
+        company = get_company()
+
+        # Create purchase user
+        User = Model.get('res.user')
+        Group = Model.get('res.group')
+        purchase_user = User()
+        purchase_user.name = 'Purchase'
+        purchase_user.login = 'purchase'
+        purchase_group, = Group.find([('name', '=', 'Purchase')])
+        purchase_user.groups.append(purchase_group)
+        stock_group, = Group.find([('name', '=', 'Stock')])
+        purchase_user.groups.append(stock_group)
+        purchase_user.save()
+
+        # Create account user
+        account_user = User()
+        account_user.name = 'Account'
+        account_user.login = 'account'
+        account_user.save()
+
+        # Create fiscal year
+        fiscalyear = set_fiscalyear_invoice_sequences(
+            create_fiscalyear(company))
+        fiscalyear.click('create_period')
+
+        # Create chart of accounts
+        _ = create_chart(company)
+        accounts = get_accounts(company)
+        revenue = accounts['revenue']
+        expense = accounts['expense']
+
+        # Create parties
+        Party = Model.get('party.party')
+        supplier = Party(name='Supplier')
+        supplier.save()
+
+        # Create account category
+        ProductCategory = Model.get('product.category')
+        account_category = ProductCategory(name="Account Category")
+        account_category.accounting = True
+        account_category.account_expense = expense
+        account_category.account_revenue = revenue
+        account_category.save()
+
+        # Create product
+        ProductUom = Model.get('product.uom')
+        unit, = ProductUom.find([('name', '=', 'Unit')])
+        ProductTemplate = Model.get('product.template')
+        template = ProductTemplate()
+        template.name = 'product'
+        template.default_uom = unit
+        template.type = 'goods'
+        template.purchasable = True
+        template.list_price = Decimal('10')
+        template.account_category = account_category
+        template.save()
```

Comment:
cost_method should be set to 'average' and when the goods are received ensure that the product has the correct cost_price.

The reason is that we may set the unit_price of the move correctly but at the wrong time and the value may not be updated appropriately.